src/process/shell.cr

```diff
@@ -0,0 +1,105 @@
+class Process
+  # Converts a sequence of strings to one joined string with each argument shell-quoted.
+  #
+  # This is then safe to pass as part of the command when using `shell: true` or `system()`.
+  #
+  # NOTE: The actual return value is system-dependent, so it mustn't be relied on in other contexts.
+  # See also: `quote_posix`.
+  #
+  # ```
+  # files = ["my file.txt", "another.txt"]
+  # `grep -E 'fo+' -- #{Process.quote(files)}`
+  # ```
+  def self.quote(args : Enumerable(String)) : String
+    {% if flag?(:win32) %}
+      quote_windows(args)
+    {% else %}
+      quote_posix(args)
+    {% end %}
+  end
+
+  # Shell-quotes one item, same as `quote({arg})`.
+  def self.quote(arg : String) : String
+    quote({arg})
+  end
+
+  # Converts a sequence of strings to one joined string with each argument shell-quoted.
+  #
+  # This is then safe to pass to a POSIX shell.
+  #
+  # ```
+  # files = ["my file.txt", "another.txt"]
+  # Process.quote_posix(files) # => "'my file.txt' another.txt"
+  # ```
+  def self.quote_posix(args : Enumerable(String)) : String
+    args.join(' ') do |arg|
+      if arg.empty?
+        "''"
+      elsif arg.matches? %r([^a-zA-Z0-9%+,\-./:=@_]) # not all characters are safe, needs quoting
+        "'" + arg.gsub("'", %('"'"')) + "'"          # %(foo'ba#r) becomes %('foo'"'"'ba#r')
+      else
+        arg
+      end
+    end
+  end
+
+  # Shell-quotes one item, same as `quote_posix({arg})`.
+  def self.quote_posix(arg : String) : String
+    quote_posix({arg})
+  end
+
+  # :nodoc:
+  #
+  # Converts a sequence of strings to one joined string with each argument shell-quoted.
+  #
+  # This is then safe to pass Windows API CreateProcess.
+  #
+  # NOTE: This is **not** safe to pass to the CMD shell.
+  #
+  # ```
+  # files = ["my file.txt", "another.txt"]
+  # Process.quote_windows(files) # => %("my file.txt" another.txt)
+  # ```
+  def self.quote_windows(args : Enumerable(String)) : String
+    String.build { |io| quote_windows(io, args) }
+  end
+
+  private def self.quote_windows(io : IO, args)
+    args.join(' ', io) do |arg|
```

Comment:
Aaaa I missed the argument swap

Reply:
Should we fail CI on warnings? Or maybe we already do on non-Windows

Reply:
No because there are tests that check that the deprecated methods still work.

Reply:
But we `--exclude-warnings spec/std`. Maybe it's not 100% reliable though.

Reply:
Yes, this should be improved. Not sure if it's not reliable or just not configured properly...

Reply:
I think that the only reliable way is to fail on warnings on std_spec, and add deprecated_spec target were deprecations can happen. That way we can error on std_spec.

Reply:
Handled in #9369